### PR TITLE
[Crypto] fix memory leak in unhappy path

### PR DESCRIPTION
**Pull request by tarakby**
- fix memory leak for some unhappy paths using a similar approach as linux memory management (Thanks @SaveTheRbtz for the suggestion)
 - minor optimization minimizing allocations of memory.
 - check malloc and calloc output to abort smoothly in case of failures.

### CODE REVIEW

File: crypto/bls_core.c [hunk 1]
```diff
@@ -341,51 +372,68 @@ typedef struct st_node {
...
     // build the tree in a top-down way
     t->left = build_tree(left_len, &pks[0], &sigs[0]);
+    if (!t->left) { free_tree(t); return NULL; }
+
     t->right = build_tree(right_len, &pks[left_len], &sigs[left_len]);
+    if (!t->right) { free_tree(t); return NULL; }
     // sum the children
     ep_add_jacob(t->sig, t->left->sig, t->right->sig);
     ep2_add_projc(t->pk, t->left->pk, t->right->pk); 
     return t;
```
Comment (SaveTheRbtz) [inline crypto/bls_core.c:434]: You may consider adding a separate error handling trailer after the `return t;`:
```
err:
    free(new_sig);
errSig:
    free(new_pk);
errPk:
    return NULL;
```

that way we'll be slightly more consistent.

**Review by SaveTheRbtz — APPROVED** [review]

**Review by SaveTheRbtz — APPROVED** [review]

**Review by durkmurder — APPROVED** [review]
LGTM

**Comment by tarakby** [comment]
bors merge

**Comment by tarakby** [comment]
bors retry

**Pull request merged**